Xext/xvdisp.c

```diff
@@ -364,46 +127,64 @@ ProcXvQueryAdaptors(ClientPtr client)
         pa++;
     }
 
+    char *payload = calloc(1, totalSize);
+    if (!payload)
+        return BadAlloc;
+    char *walk = payload;
+
     rep.length = bytes_to_int32(totalSize);
 
-    _WriteQueryAdaptorsReply(client, &rep);
+    if (client->swapped) {
+        swaps(&rep.sequenceNumber);
+        swapl(&rep.length);
+        swaps(&rep.num_adaptors);
+    }
 
     na = pxvs->nAdaptors;
     pa = pxvs->pAdaptors;
     while (na--) {
+        xvAdaptorInfo *ainfo = (xvAdaptorInfo*)walk;
+
+        ainfo->base_id = pa->base_id;
+        ainfo->num_ports = pa->nPorts;
+        ainfo->type = pa->type;
+        ainfo->name_size = nameSize = strlen(pa->name);
+        ainfo->num_formats = pa->nFormats;
+
+        if (client->swapped) {
+            swapl(&ainfo->base_id);
+            swaps(&ainfo->name_size);
+            swaps(&ainfo->num_ports);
+            swaps(&ainfo->num_formats);
+        }
 
-        ainfo.base_id = pa->base_id;
-        ainfo.num_ports = pa->nPorts;
-        ainfo.type = pa->type;
-        ainfo.name_size = nameSize = strlen(pa->name);
-        ainfo.num_formats = pa->nFormats;
-
-        _WriteAdaptorInfo(client, &ainfo);
-
-        WriteToClient(client, nameSize, pa->name);
+        walk += sizeof(ainfo);
+        memcpy(walk, pa->name, nameSize);
+        walk += pad_to_int32(nameSize);
 
         nf = pa->nFormats;
         pf = pa->pFormats;
         while (nf--) {
-            format.depth = pf->depth;
-            format.visual = pf->visual;
-            _WriteFormat(client, &format);
+            xvFormat *format = (xvFormat *)walk;
+            format->depth = pf->depth;
+            format->visual = pf->visual;
+            if (client->swapped) swapl(&format->visual);
```

Comment:
Shouldn’t walk be moved forward after this to avoid writing multiple times to the same place?

Reply:
This and the alignment issue is exactly why im not in favor of this type of buffer writing with walking. It's very bug-prone.

Reply:
@dec05eba feel free to propose a better solution.
Note, that's all just preparation for this one: https://github.com/X11Libre/xserver/tree/wip/swapping_new